python/ray/train/mosaic/_mosaic_utils.py

```diff
@@ -0,0 +1,54 @@
+from typing import Any, Dict, Optional, List
+import torch
+
+from composer.loggers import Logger
+from composer.loggers.logger_destination import LoggerDestination
+from composer.core.state import State
+
+from ray.air import session
+
+
+class RayLogger(LoggerDestination):
+    """A logger to relay information logged by composer models to ray.
+
+    This logger allows utilizing all necessary logging and logged data handling provided
+    by the Composer library. All the logged information is saved in the data dictionary
+    every time a new information is logged, but to reduce unnecessary reporting, the
+    most up-to-date logged information is reported as metrics every batch checkpoint and
+    epoch checkpoint (see Composer's Event module for more details).
+
+    Because ray's metric dataframe will not include new keys that is reported after the
+    very first report call, any logged information with the keys not included in the
+    first batch checkpoint would not be retrievable after training. In other words, if
+    the log level is greater than `LogLevel.BATCH` for some data, they would not be
+    present in `Result.metrics_dataframe`. To allow preserving those information, the
+    user can provide keys to be always included in the reported data by using `keys`
+    argument in the constructor. For `MosaicTrainer`, use
+    `trainer_init_config['log_keys']` to populate these keys.
+
+    Note that in the Event callback functions, we remove unused variables, as this is
+    practiced in Mosaic's composer library.
+
+    Args:
+        keys: the key values that will be included in the reported metrics.
+    """
+
+    def __init__(self, keys: List[str] = None) -> None:
+        self.data = {}
+        if keys:
+            for key in keys:
+                self.data[key] = None
+
+    def log_metrics(self, metrics: Dict[str, Any], step: Optional[int] = None) -> None:
+        self.data.update(metrics.items())
+        for key, val in self.data.items():
+            if isinstance(val, torch.Tensor):
+                self.data[key] = val.item()
+
+    def batch_checkpoint(self, state: State, logger: Logger) -> None:
+        del logger  # unused
+        session.report(self.data)
+
+    def epoch_checkpoint(self, state: State, logger: Logger) -> None:
+        del logger  # unused
+        session.report(self.data)
```

Comment:
We shouldn't report on at both the batch level and the epoch level. Each call to `session.report` should be 1 iteration, so if we log at both, we will be double counting.

For now, I would say let's just log only at every epoch. We can see in the future if we want to give users the ability to configure this.

Reply:
I think we'll definitely have users want to do it on either level - that was the case with HF, where we started with epochs only and had to add steps too.

Reply:
Completely agree @Yard1. I’m thinking we can default to epoch for now and then add batch support in a follow up.

Reply:
This has been updated!
We are reporting each epoch now, but we also report after the fit call, in case the training ends before an epoch checkpoint call could be made. This adds an extra report call, in which an epoch checkpoint can be double counted.  -- but we can also make it so that this last call is made only if there are extra batch runs after the last epoch run.

Reply:
The mentioned change has been applied.